Prepend task id if it exists to "ray task list --detail"

ProjectsByJackHe opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

This change addresses this issue: https://github.com/ray-project/ray/issues/30805.

It solves a UX issue where the "task_id" field isn't printed FIRST. So this change prepends it, which can be shown in the following screenshot: 

UPDATED screenshot:

<img width="540" alt="Screen Shot 2023-02-21 at 4 08 12 PM" src="https://user-images.githubusercontent.com/43735701/220487130-d69b2f9c-1d4b-459d-81be-8750cbf87936.png">


<img width="485" alt="Screen Shot 2023-02-13 at 5 54 54 PM" src="https://user-images.githubusercontent.com/43735701/218618474-8b973c91-42b9-42a4-a432-4821f9afbffc.png">


## Related issue number

Closes Part 2 of #30805 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

scottsun94 commented:
Thanks for working on this! cc: @rkooo567 @rickyyx

rickyyx commented:
Hey @ProjectsByJackHe thanks for contributing!

Adding the task id as a separate element that's at the same level as other data might not be that intuitive. 

How about we change each field from 
```
[
    {
        task_id: xxxx,
        ...
        other data: yyy
    },
    ...
]
```

to something like 

```
[
    {
        xxxx: {
            ...
            other data: yyy
        }
    },
    ....
]
```


This should produce yaml of: 
```
- <task_id> 
   - actor_id: xxx
   - others....

```

ProjectsByJackHe commented:
> Hey @ProjectsByJackHe thanks for contributing!
> 
> Adding the task id as a separate element that's at the same level as other data might not be that intuitive.
> 
> How about we change each field from
> 
> ```
> [
>     {
>         task_id: xxxx,
>         ...
>         other data: yyy
>     },
>     ...
> ]
> ```
> 
> to something like
> 
> ```
> [
>     {
>         xxxx: {
>             ...
>             other data: yyy
>         }
>     },
>     ....
> ]
> ```
> 
> This should produce yaml of:
> 
> ```
> - <task_id> 
>    - actor_id: xxx
>    - others....
> ```
> 
> Wdyt ? @scottsun94 @ProjectsByJackHe

Yeah, that makes a lot of sense! I can also prepend "task_id: " to xxxx. Change pushed :)

ProjectsByJackHe commented:
> > Hey @ProjectsByJackHe thanks for contributing!
> > Adding the task id as a separate element that's at the same level as other data might not be that intuitive.
> > How about we change each field from
> > ```
> > [
> >     {
> >         task_id: xxxx,
> >         ...
> >         other data: yyy
> >     },
> >     ...
> > ]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > to something like
> > ```
> > [
> >     {
> >         xxxx: {
> >             ...
> >             other data: yyy
> >         }
> >     },
> >     ....
> > ]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This should produce yaml of:
> > ```
> > - <task_id> 
> >    - actor_id: xxx
> >    - others....
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Wdyt ? @scottsun94 @ProjectsByJackHe
> 
> Yeah, that makes a lot of sense! I can also prepend "task_id: " to xxxx. Change pushed :)

@rickyyx

rickyyx commented on python/ray/experimental/state/state_cli.py:
```diff
@@ -173,7 +173,15 @@ def output_with_format(
+            {
+                ("task_id: " + state["task_id"])
+                if "task_id" in state
+                else "No task_id found": state
```
we should always have task id - so this is probably not needed.

ProjectsByJackHe replied:
sounds good! I can remove.

rickyyx commented on python/ray/experimental/state/state_cli.py:
```diff
@@ -173,7 +173,15 @@ def output_with_format(
-        return yaml.dump(state_data, indent=4, explicit_start=True)
+        augmented_state_data = [
+            {
+                ("task_id: " + state["task_id"])
```
I guess we might want to use the extra task id as the key for this dict?

ProjectsByJackHe replied:
hmm, I am not sure what you mean, are you saying we should pull out the state[task_id] into a separate local variable to re-use elsewhere?

rickyyx reviewed (commented):
Hey @ProjectsByJackHe thanks for the quick iteration. Could you add the manual testing results (like what you did in the original description)?

ProjectsByJackHe commented:
> Hey @ProjectsByJackHe thanks for the quick iteration. Could you add the manual testing results (like what you did in the original description)?

Updated! And I also updated the PR to remove the redundant check for existence of "task_id."

rickyyx commented:
Ah, seems some tests were broken. e.g. https://buildkite.com/ray-project/oss-ci-build-pr/builds/12198#01864daa-14e9-4732-8bbb-dee239e97b7a

I think the reason is we are also passing states info other than task state into this function `output_with_format`. For example, we will also call this here for logs:
https://github.com/ray-project/ray/blob/e78def4c701f934b094f1a50dbb2e5e11731a399/python/ray/experimental/state/state_cli.py#L949


So I guess we will need to make sure we only do what you did in this PR at the right place for the right schema. Sorry for not pointing it out earlier. Maybe we could check if the schema is a TaskSchema here instead, so we only do the task id prepending if it's a task state:
https://github.com/ray-project/ray/blob/e78def4c701f934b094f1a50dbb2e5e11731a399/python/ray/experimental/state/state_cli.py#L278-L288


Btw, you could run commands like `pytest -vv python/ray/tests/test_state_api.py` in the project root to do local testing for verification.

ProjectsByJackHe commented:
> Ah, seems some tests were broken. e.g. https://buildkite.com/ray-project/oss-ci-build-pr/builds/12198#01864daa-14e9-4732-8bbb-dee239e97b7a
> 
> I think the reason is we are also passing states info other than task state into this function `output_with_format`. For example, we will also call this here for logs:
> 
> https://github.com/ray-project/ray/blob/e78def4c701f934b094f1a50dbb2e5e11731a399/python/ray/experimental/state/state_cli.py#L949
> 
> So I guess we will need to make sure we only do what you did in this PR at the right place for the right schema. Sorry for not pointing it out earlier. Maybe we could check if the schema is a TaskSchema here instead, so we only do the task id prepending if it's a task state:
> 
> https://github.com/ray-project/ray/blob/e78def4c701f934b094f1a50dbb2e5e11731a399/python/ray/experimental/state/state_cli.py#L278-L288
> 
> Btw, you could run commands like `pytest -vv python/ray/tests/test_state_api.py` in the project root to do local testing for verification.

Sounds good, thanks for pointing this out! Changes pushed.

rickyyx reviewed (approved):
Awesome, thanks! 

Could you add some screenshots to the PR to validate these changes below? Since I believe its not covered by unittesting.

ProjectsByJackHe commented:
> Awesome, thanks!
> 
> Could you add some screenshots to the PR to validate these changes below? Since I believe its not covered by unittesting.

Sure, added to PR description the following screenshot:
<img width="535" alt="Screen Shot 2023-02-21 at 4 08 51 PM" src="https://user-images.githubusercontent.com/43735701/220487207-f3c9c362-2a5a-418a-95c5-08847a8da317.png">

rickyyx commented:
cc @rkooo567

rkooo567 commented:
Thanks for the contribution!

cadedaniel commented:
Nice @ProjectsByJackHe ! :)
